src/core/system_parameters/beidou_cnav2_navigation_message.cc

```diff
@@ -0,0 +1,794 @@
+ * \see <a href="http://m.beidou.gov.cn/xt/gfxz/201712/P020171226742357364174.pdf">BEIDOU ICD</a>
+ *
+ * -------------------------------------------------------------------------
+ *
+ * Copyright (C) 2010-2015  (see AUTHORS file for a list of contributors)
+ *
+ * GNSS-SDR is a software defined Global Navigation
+ *          Satellite Systems receiver
+ *
+ * This file is part of GNSS-SDR.
+ *
+ * GNSS-SDR is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * GNSS-SDR is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GNSS-SDR. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * -------------------------------------------------------------------------
+ */
+
+#include "beidou_cnav2_navigation_message.h"
+#include "gnss_satellite.h"
+#include <glog/logging.h>
+
+
+void Beidou_Cnav2_Navigation_Message::reset()
+{
+    // Satellite Identification
+    i_satellite_PRN = 0;
+    i_alm_satellite_slot_number = 0;  //!< SV Orbit Slot Number
+    flag_update_slot_number = false;
+
+    // Ephemeris Flags
+    flag_all_ephemeris = false;
+    flag_ephemeris_str_1 = false;
+    flag_ephemeris_str_2 = false;
+
+    // Almanac Flags
+    flag_almanac_str_31 = false;
+
+    // UTC and System Clocks Flags
+    flag_utc_model_valid = false;   //!< If set, it indicates that the UTC model parameters are filled
+    flag_utc_model_str_5 = false;   //!< Clock info send in string 5 of navigation data
+    flag_utc_model_str_15 = false;  //!< Clock info send in string 15 of frame 5 of navigation data
+
+    // broadcast orbit 1
+    flag_TOW_set = false;
+    flag_TOW_new = false;
+
+    flag_CRC_test = false;
+    d_frame_ID = 0;
+    d_string_ID = 0;
+    i_channel_ID = 0;
+
+    // Clock terms
+    d_satClkCorr = 0.0;
+    d_dtr = 0.0;
+    d_satClkDrift = 0.0;
+
+    // Data update information
+    d_previous_tb = 0.0;
+    for (unsigned int i = 0; i < BEIDOU_CA_NBR_SATS; i++)
+        d_previous_Na[i] = 0.0;
+
+    std::map<int, std::string> satelliteBlock;  //!< Map that stores to which block the PRN belongs http://www.navcen.uscg.gov/?Do=constellationStatus
+
+    auto gnss_sat = Gnss_Satellite();
+    std::string _system("BEIDOU");
+    //TODO SHould number of channels be hardcoded?
+    for (unsigned int i = 1; i < 14; i++)
+        {
+            satelliteBlock[i] = gnss_sat.what_block(_system, i);
+        }
+}
+
+
+Beidou_Cnav2_Navigation_Message::Beidou_Cnav2_Navigation_Message()
+{
+    reset();
+}
+
+
+bool Beidou_Cnav2_Navigation_Message::CRC_test(std::bitset<BEIDOU_CNAV2_STRING_BITS> bits)
```

Comment:
This CRC code seems awful similar to that of GLONASS, Are you sure this is right?